Support new float:round langlib function

KavinduZoysa opened the pull request:
## Purpose

Fixes #35543

gimantha commented on langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/Round.java:
```diff
@@ -31,7 +34,46 @@
+            // Therefore, if `fractionDigits` is very large, `x` will not be changed.
+            return x;
+        }
+        if (fractionDigits < Integer.MIN_VALUE) {
```
Can fractalDigits be less than 0?

KavinduZoysa replied:
Yes please check https://github.com/ballerina-platform/ballerina-spec/issues/1074

KRVPerera commented on langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/Round.java:
```diff
@@ -31,7 +34,46 @@
+        BigDecimal xTmp = xInBigDecimal;
+        int digitsTmp = fractionDigitsAsInt;
+        if (digitsTmp > 0) {
+            while (digitsTmp-- > 0) {
```
Shall we add some comments to explain this logic in high level. Looks to me like if we multiply by 100 this will go faster.
BigDecimal has a `scale` method to know decimal points in advance. So we can check if 
`x.scale() <= digitsTmp` in one line

KRVPerera commented on langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/Round.java:
```diff
@@ -31,7 +34,46 @@
+                xTmp = xTmp.multiply(BigDecimal.TEN);
+            }
+        } else {
+            while (digitsTmp++ < 0) {
```
https://stackoverflow.com/questions/35792590/how-to-check-number-of-digits-from-bigdecimal
There is a trick to this as well..
```Java
xInBigDecimal = xInBigDecimal.stripTrailingZeros();
int  y = xInBigDecimal.precision() - xInBigDecimal.scale()
if (y > digitsTmp) {
  return 0;
}
```

KRVPerera reviewed (commented):
LGTM except for two loops written to get a numerical results.
There are easy ways to get them without looping and added them as comments.

KavinduZoysa commented:
> LGTM except for two loops written to get a numerical results. There are easy ways to get them without looping and added them as comments.

changed the logic please check

SandaruJayawardana reviewed (dismissed):
LGTM

MaryamZi commented on langlib/lang.float/src/main/ballerina/float.bal:
```diff
@@ -106,19 +106,22 @@ public isolated function abs(float x) returns float = @java:Method {
-#
-# Returns the float value that is a mathematical integer and closest to parameter `x`.
-# If there are two such values, choose the one that is even
+# Round a float value to a specified number of digits.
```
```suggestion
# Rounds a float value to a specified number of digits.
#
```

KavinduZoysa replied:
Changed

MaryamZi commented on langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/Round.java:
```diff
@@ -31,7 +34,39 @@
+        return xInBigDecimal.setScale(fractionDigitsAsInt, RoundingMode.HALF_EVEN).doubleValue();
+    }
+
+    private static boolean isSpecialValue(double x) {
+        return Double.isInfinite(x) || Double.isNaN(x) || x == 0.0d;
```
In this particular case, since "special value" doesn't quite convey what we're checking for and it isn't a term from the spec, I think we can just inline the check.

KavinduZoysa replied:
Changed

MaryamZi commented on langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/Round.java:
```diff
@@ -31,7 +34,39 @@
+            if (fractionDigitsAsInt > scale) {
+                return x;
+            }
+        } else {
+            if (-fractionDigitsAsInt > (xInBigDecimal.precision() - scale)) {
```
Can be merged into an else if?

KavinduZoysa replied:
Yes, done

MaryamZi commented on langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibFloatTest.java:
```diff
@@ -117,5 +117,17 @@ public void testFromString(String function) {
-    }  
+    }
+
+    @Test(dataProvider = "functionsToTestRound")
```
Why do we need a data provider if we have only one test function?

MaryamZi commented on langlib/langlib-test/src/test/resources/test-src/floatlib_test.bal:
```diff
@@ -415,6 +415,176 @@ function testFromStringNegative() {
 }
 
+function testRound() {
+    assertEquality(float:round(123.123f, 2), 123.12f);
```
Let's add a few tests with the method call syntax also.

MaryamZi replied:
Also with named arguments.

KavinduZoysa replied:
Added

MaryamZi commented on langlib/langlib-test/src/test/resources/test-src/floatlib_test.bal:
```diff
@@ -415,6 +415,176 @@ function testFromStringNegative() {
 }
 
+function testRound() {
+    assertEquality(float:round(123.123f, 2), 123.12f);
```
The float suffix isn't required in these tests?

KavinduZoysa replied:
changed

MaryamZi commented on langlib/lang.float/src/main/ballerina/float.bal:
```diff
@@ -106,19 +106,22 @@ public isolated function abs(float x) returns float = @java:Method {
-#
-# Returns the float value that is a mathematical integer and closest to parameter `x`.
-# If there are two such values, choose the one that is even
+# Rounds a float value to a specified number of digits.
```
```suggestion
# Rounds a float value to a specified number of digits.
#
```

KavinduZoysa replied:
Added
